### ISSUE TITLE
Minor Release 3.6.0 - Scheduled for 14. July 2022

### ISSUE THREAD

**Issue by boaks** [eclipse/californium#2040]
For details, please see [3.6.0](https://projects.eclipse.org/projects/iot.californium/releases/3.6.0)
# Please provide basic issue information

Issues may be answered much faster, if you provide more details ahead!

In the last years, especially two kind of issues consumed a lot of our time, and it will be really much faster, if you provide therefore this information **ahead**.

## Issue type 1:

Californium is used together with an other CoAP/DTLS implementation. Please provide, where you run Californium (client or server side) and which other implementation you use. Even, if the other implementation is a proprietary one, it is important to know, that it is not Californium. Without traces and captures in the very most cases we can not help you. So please really consider to provide that all ahead! And, please captures, no pictures!

## Issue type 2:

Network environment, e.g. k8s, docker, NATs, firewalls, virtual machines. All that may apply unaware changes to your traffic and may result in failures. In some cases it is also interesting to other users, if such issues occurs. So, still please report them. But the most of them can not be fixed in Californium, they must be fixed in that network environment. Therefore it is very important, that you list all the used components. In order to ensure, that it is no Californium issue, we kindly ask you to verify, that the issue doesn't occur, if a simple network environment is used (e.g. running in two raspberry PIs).

## Information ahead to speed up the processing of issues:

* branch/release your faced the issue. Currently
    * (1.0 is a previous release, usually no bugfixes are applied)
    * (1.1 is the previous master, no plans to continue this)
    * 2.7.x is the last previous major release
    * and main is the development branch include the current major release 3.x.y
* If you use a SNAPSHOT, ensure your updated to the latest. Please provide the commit your using
* If your issue is related to a combination with other implementations, please mention that implementation (including proprietary ones). And the sides your using it, client or server.
* If your issue is related to your own snippet, please provide that
* If you have logs, please provide them
* If you have tcpdump captures (wireshark), please provide them. Screenshot of wireshark (or even movies) don't help. The capture itself is required in order to explore the messages.

(See https://github.com/eclipse/californium/wiki/Logs-and-IP-Capturing-%E2%80%90-How-To-Provide-The-Right-Information .)

**Comment by boaks** [comment]
The 3.6.0 is released and available on [Maven Central](https://search.maven.org/search?q=g:org.eclipse.californium%20a:parent%20v:3.6.0) and the [Eclipse Repository](https://repo.eclipse.org/content/repositories/californium-releases/org/eclipse/californium).

The tools and actinium are only available on the [Eclipse Repository](https://repo.eclipse.org/content/repositories/californium-releases/org/eclipse/californium).

**Comment by boaks** [comment]
> Just to let know that, there is something which looks like a behavior break with this version.

Thanks!
FMPOV, it breaks a "misbehavior" ;-).

> This could raise exception at unexpected moment for people which was using previous 3.x version

I don't get this. This adds an additional check to `DtlsConnectorConfig.builder.build()` and works, as all the others there. AFAIK, the only affected use-cases are tests with broken key-pairs, and for them, you may disable it.

**Comment by boaks** [comment]
@sbernard31 

Any news/feedback on the reported  API break? 

Did you find it on tests with mismatching public/private keys?

Or have I overseen use-cases?

**Comment by sbernard31** [comment]
**Short version**

I raise this just in case this was not done on purpose. 
From Leshan v2.0.0-Mx, not a big deal, I just go back to previous behavoir in Leshan Client. 

In all case, I strongly advice to remove the `ex.printStackTrace();`.

**Long answer**

In a general way : 
I just said that changing this kind of behavior could create "unexpected problem".
If the idea of semantic versioning is to ensure (as far as possible)  backward compatibility, my opinion is that this kind of changes should be avoid  for minor version : 
- new minor : you add new/better behavior without changing default behavior
- new major : you change default behavior.

Anyway, I understand why this could also be considered  as a bug/ or miss behavior fix.

More concretely for Leshan integration : 
 1. For leshan v2.0.0 (in development),   don't care if there is  API break in californium before the official v2.0.0 release. I have no problem integrating new major.
 2. mismatching public/private keys tests failed. (Not so easy to adapt the code but this is life)
 3. In Leshan client, credentials could change dynamically (bootstrap).  Before bad key make handshake failed, then device try to bootstrap again. With new behavior, exception is raised and client crash/shutdown.
 
 So for now I just go back to previous behavior in Leshan Client.
 I guess in Leshan Client this kind of check  should be done at bootstrap time in bootstrap config consistency checker. (not done for now)

 About 3., I guess we can consider this is a bad behavior of Leshan because this kind of exception should be handled correctly but this is what I called "unexpected problem".

I don't try to convince you, I just give more details as you asked.

**Comment by boaks** [comment]
> I don't try to convince you, I just give more details as you asked.

Thanks! The `ex.printStackTrace()` is a overseen left over.

> In Leshan client, credentials could change dynamically (bootstrap). Before bad key make handshake failed, then device try to bootstrap again. With new behavior, exception is raised and client crash/shutdown.

Generally, the checks in `DtlsConnectorConfig.Builder.build()` have been improved over the time. And the last is that detection of private/public key mismatches. I'm not sure, if "misconfiguration" is considered, not dealing with the `RuntimeException` is really a good solution. Maybe, for midterm, it's easier to "fallback" also on such an configuration exception.

**Comment by boaks** [comment]
:exclamation::exclamation::exclamation: **Important Note:** :exclamation::exclamation::exclamation:

This bugfix is required for all users of Californium 3.0.0 - 3.5.0, 
which are using DTLS resumption and DTLS_VERIFY_PEERS_ON_RESUMPTION_THRESHOLD values larger than 0!
It provides the fix for 

[CVE-2022-2576](https://cve.report/CVE-2022-2576)

**Comment by jvermillard** [comment]
On my local machine (openjdk 17), leshan tests complains about the key (the exact issue pointed by Simon).

But I struggle to understand why, the key is generated like this:

https://github.com/eclipse/leshan/blob/master/leshan-server-cf/src/test/java/org/eclipse/leshan/server/californium/bootstrap/LeshanBootstrapServerBuilderTest.java#L61-L85

**Comment by jvermillard** [comment]
I guess the starting point is broken and we should generate one properly

**Comment by boaks** [comment]
Some weeks/months ago, I recognized, that people not common with x509 mix up the private and public key firm the certificates. And then report mystic errors. e.g. server's private key, CA's public key.
With that I started to analyze, why could be done best (at least from my view point). I decided to report broken key-pairs just at the builder, as many other misconfiguration also is reported.
That even caused some of my unit tests in Californium to fail, because they use such a invalid pair to verify, that some errors are detected. With that, I added also a function to disable that check. I accidentally left the "System.out", but in the meantime this is replaced by the LOGGER and should be gone with next minor release.

So, from my side the question is:
If Californium generally checks for misconfiguration and reports the in the build by exceptions,
is it then really a good practice to "do all well and so ignore the exception"?
For me it's not. If leshan would catch the exception, it would be already clear at that point, that the bootstrap must be repeated. If it's really not want, the disable the key verification.

About the System.out, I would like to finish a first version of a [coap-to-S3-proxy](https://github.com/boaks/californium/tree/add_cloud_demo_server/demo-apps/cf-cloud-demo-server). Since yesterday the required CQs for the aws-sdk has been resolved, so I will prepare for a initial PR as work-in-progress and a feature branch. My forecast will be something short before the EclipseCon.

If you @sbernard31  or @jvermillard see any benefit in a 3.7 without the System.out earlier, let me know, then we can release the 3.7 with the S3-proxy and do that later in a 3.8.

 
If Julien or Simon see

**Comment by sbernard31** [comment]
@Achim,

Maybe @jvermillard was not clear but I guess he wanted to mention that he faces this issue but in a particular case : 
with openjdk 17 only (not previous one) and for which was considered until now as valid key pair (at least it works with previous version of JDK) See https://github.com/eclipse/leshan/issues/1298.

(this is not a complain about regression or something like this but more asking for help just in case you understand / face same kind of issue) 

I feel that a dedicated issue should have been created because this 3.6.0 annonce issue is probably not the best place.
If you want to discuss about it, I guess you can use the leshan issue. 

> If you @sbernard31 or @jvermillard see any benefit in a 3.7 without the System.out earlier, let me know

No urgency for me.

**Comment by boaks** [comment]
Is it possible for you to share the affected keys (demo-keys?)? 
Or at least the public-key and certificate?

**Comment by sbernard31** [comment]
This is just RPK (no certificate) and it was created like this : https://github.com/eclipse/leshan/blob/master/leshan-server-cf/src/test/java/org/eclipse/leshan/server/californium/bootstrap/LeshanBootstrapServerBuilderTest.java#L61-L85

**Comment by boaks** [comment]
> This is just RPK (no certificate) and it was created like this

And running that with java 11 succeeds and fails with 17?

**Comment by sbernard31** [comment]
Yep :shrug:

**Comment by sbernard31** [comment]
> Just to mention:
> 
> [CertificateConfigurationHelper](https://github.com/eclipse/californium/blob/main/scandium-core/src/main/java/org/eclipse/californium/scandium/dtls/x509/CertificateConfigurationHelper.java#L291-L340)
> 
> does the check. I signs with the private key and verifies with the public key.

Yep I looked at this but don't get why JDK behave differently for this particular key pairs :thinking:

**Comment by boaks** [comment]
I can reproduce the behavior (I added a test for the leshan keys to CertificateConfigurationHelperTest.
Works with java11, fails with java17 ...
Yes, Oracle changed a lot in the EC implementation, you may remember [ECDSA - CVE-2022-21449](https://cve.mitre.org/cgi-bin/cvename.cgi?name=CVE-2022-21449).
I will check that. But it may be not too easy. And I'm not sure, if the key-pair works, but the verification fails. Or it doesn't work.

**Comment by sbernard31** [comment]
I checked the pair with : 
- https://kjur.github.io/jsrsasign/sample/sample-ecdsa.html  (written in javascrip so at least this is not JDK behind)
- https://www.javacardos.com/tools/ecdsa-sign-verify

And both seems to work with this key pair. :thinking:

**Comment by boaks** [comment]
OK, my result so far:
All your keys have the high-bit set and may be interpreted as negative numbers. But EC is based on positive numbers.
If you add a "00" at the head, ... it works ...

**Comment by boaks** [comment]
I hope, it works with the 00 for you as well.
Just as cross check:
If you now switch off the check and try a handshake with the original key, does that succeed?

**Comment by sbernard31** [comment]
> I hope, it works with the 00 for you as well.

It seems to work.

>  If you now switch off the check and try a handshake with the original key, does that succeed?

My first test seems to show it works :grimacing:, I guess it should not, or  ?

**Comment by boaks** [comment]
> My first test seems to show it works grimacing, I guess it should not, or ?

I would have assumed **not**. 
Anyway, in the real handshake the public key is sent and processed by the other peer. 
The other peer load that keys from ASN.1, which may fix that on the fly :-):

**Comment by boaks** [comment]
From my side:
I don't think, that the verify introduces a real issue.
At least, if it's assumed, that the public key do not only hold the right bytes, but are applicable as well.

**Comment by sbernard31** [comment]
> I also tried with java 15, which has also the new EC implementation and it fails with that.

I guess this confirms the idea : https://bugs.openjdk.org/browse/JDK-8183666

**Comment by boaks** [comment]
See PR #2058

**Comment by boaks** [comment]
Yes, once on the right trail, the information starts to give the right picture.

(I still think, that this verification is a good idea, even if I learned now, that there are some pitfalls.)

**Comment by boaks** [comment]
You're welcome.
